File: Problem 014/solution.py
def collatzSequenceUtil(n, collLenDict):
    if n in collLenDict:
        return collLenDict[n]

    if n == 1:
        collLenDict[n] = 1

    elif n % 2 == 0:
        collLenDict[n] = 1 + collatzSequenceUtil(n // 2, collLenDict)
    else:
        collLenDict[n] = 1 + collatzSequenceUtil((3 * n) + 1, collLenDict)

    return collLenDict[n]


def collatzSequence(n):
    lenght = -1
    num = 0
    collLenDict = {}

    collatzSequenceUtil(n, collLenDict)

    for i in range(1, n):
        if i not in collLenDict:
            collatzSequenceUtil(i, collLenDict)

        currentLen = collLenDict[i]

        if lenght < currentLen:
            lenght = currentLen
            num = i

    return num, lenght


print(collatzSequence(1000000))

File: Problem 014/solution.go
package main

import "fmt"

func collatzSequenceUtil(n int, collLenMap map[int]int) int {
	if _, ok := collLenMap[n]; ok {
		return collLenMap[n]
	}

	if n == 1 {
		collLenMap[n] = 1
	} else if n%2 == 0 {
		collLenMap[n] = 1 + collatzSequenceUtil(n/2, collLenMap)
	} else {
		collLenMap[n] = 1 + collatzSequenceUtil((3*n)+1, collLenMap)
	}

	return collLenMap[n]
}

func collatzSequence(n int) (int, int) {
	lenght := -1
	num := 0
	collLenMap := make(map[int]int)

	collatzSequenceUtil(n, collLenMap)

	for i := 1; i < n; i++ {
		if _, ok := collLenMap[i]; !ok {
			collatzSequenceUtil(i, collLenMap)
		}

		currentLen := collLenMap[i]

		if lenght < currentLen {
			lenght = currentLen
			num = i
		}
	}

	return num, lenght
}

func main() {
	fmt.Println(collatzSequence(1000000))
}
